File: common/src/main/java/com/like/common/view/BadgeView.java
package com.like.common.view;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.FrameLayout.LayoutParams;
import android.widget.TabWidget;

public class BadgeView extends androidx.appcompat.widget.AppCompatTextView {
    private static int dpRadius = 1000;
    private int count;

    public BadgeView(Context context) {
        this(context, null);
    }

    public BadgeView(Context context, AttributeSet attrs) {
        this(context, attrs, android.R.attr.textViewStyle);
    }

    public BadgeView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        if (!(getLayoutParams() instanceof LayoutParams)) {
            LayoutParams layoutParams =
                    new LayoutParams(
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                            Gravity.RIGHT | Gravity.TOP);
            setLayoutParams(layoutParams);
        }

        // set default font
        setTextColor(Color.WHITE);
        setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
        setGravity(Gravity.CENTER);

        setBackgroundColor(Color.RED);

        setBadgeCount(0);
    }

    @Override
    public void setBackgroundColor(int color) {
        int radius = dp2Px(dpRadius);
        float[] radiusArray = new float[]{radius, radius, radius, radius, radius, radius, radius, radius};

        RoundRectShape roundRect = new RoundRectShape(radiusArray, null, null);
        ShapeDrawable bgDrawable = new ShapeDrawable(roundRect);
        bgDrawable.getPaint().setColor(color);
        setBackground(bgDrawable);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int w = getMeasuredWidth();
        int h = getMeasuredHeight();
        int padding;
        if (w <= h && getText() != null && getText().length() < 2) {// 宽小于高，并且只有1个字符，就画圆
            padding = (h - w) / 2;
            if (padding < 2) {
                padding = 2;
            }
            super.setMeasuredDimension(h, h);
        } else {
            padding = dp2Px(4);
            super.setMeasuredDimension(w, h);
        }
        setPadding(padding, 0, padding, 0);
    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        if (text == null || text == "" || text.toString().equalsIgnoreCase("0")) {
            setVisibility(View.GONE);
        } else {
            setVisibility(View.VISIBLE);
        }
        super.setText(text, type);
    }

    public String transformCountToText(int count) {
        return String.valueOf(count);
    }

    public void setBadgeCount(int count) {
        this.count = count;
        setText(transformCountToText(count));
    }

    public int getBadgeCount() {
        return count;
    }

    public void setBadgeGravity(int gravity) {
        LayoutParams params = (LayoutParams) getLayoutParams();
        params.gravity = gravity;
        setLayoutParams(params);
    }

    public int getBadgeGravity() {
        LayoutParams params = (LayoutParams) getLayoutParams();
        return params.gravity;
    }

    public void setBadgeMargin(int dipMargin) {
        setBadgeMargin(dipMargin, dipMargin, dipMargin, dipMargin);
    }

    public void setBadgeMargin(int leftDipMargin, int topDipMargin, int rightDipMargin, int bottomDipMargin) {
        LayoutParams params = (LayoutParams) getLayoutParams();
        params.leftMargin = dp2Px(leftDipMargin);
        params.topMargin = dp2Px(topDipMargin);
        params.rightMargin = dp2Px(rightDipMargin);
        params.bottomMargin = dp2Px(bottomDipMargin);
        setLayoutParams(params);
    }

    public int[] getBadgeMargin() {
        LayoutParams params = (LayoutParams) getLayoutParams();
        return new int[]{params.leftMargin, params.topMargin, params.rightMargin, params.bottomMargin};
    }

    public void incrementBadgeCount(int increment) {
        setBadgeCount(increment + getBadgeCount());
    }

    public void decrementBadgeCount(int decrement) {
        setBadgeCount(getBadgeCount() - decrement);
    }

    /*
     * Attach the BadgeView to the TabWidget
     *
     * @param target the TabWidget to attach the BadgeView
     *
     * @param tabIndex index of the tab
     */
    public void setTargetView(TabWidget target, int tabIndex) {
        View tabView = target.getChildTabViewAt(tabIndex);
        setTargetView(tabView);
    }

    /*
     * Attach the BadgeView to the target dragPhotoView
     *
     * @param target the dragPhotoView to attach the BadgeView
     */
    public void setTargetView(View target) {
        if (getParent() != null) {
            ((ViewGroup) getParent()).removeView(this);
        }

        if (target == null) {
            return;
        }

        if (target.getParent() instanceof FrameLayout) {
            ((FrameLayout) target.getParent()).addView(this);

        } else if (target.getParent() instanceof ViewGroup) {
            // use a new Framelayout container for adding badge
            ViewGroup parentContainer = (ViewGroup) target.getParent();
            int groupIndex = parentContainer.indexOfChild(target);
            parentContainer.removeView(target);

            FrameLayout badgeContainer = new FrameLayout(getContext());
            ViewGroup.LayoutParams parentLayoutParams = target.getLayoutParams();

            badgeContainer.setLayoutParams(parentLayoutParams);
            target.setLayoutParams(new ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

            parentContainer.addView(badgeContainer, groupIndex, parentLayoutParams);
            badgeContainer.addView(target);

            badgeContainer.addView(this);
        } else if (target.getParent() == null) {
            Log.e(getClass().getSimpleName(), "ParentView is needed");
        }

    }

    /*
     * converts dip to px
     */
    private int dp2Px(float dip) {
        return (int) (dip * getContext().getResources().getDisplayMetrics().density + 0.5f);
    }
}

File: common/src/main/java/com/like/common/view/BadgeView.kt
package com.like.common.view

import android.content.Context
import android.graphics.*
import android.text.TextPaint
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.like.common.util.dp
import com.like.common.util.sp

/**
 * 角标
 * 注意：
 * 1、如果用代码创建对象，则需要调用[setTargetView]方法。
 * 2、如果在xml中直接使用，则不需要调用[setTargetView]方法。
 */
open class BadgeView(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {
    private val textPaint by lazy {
        TextPaint().apply {
            isAntiAlias = true
            color = Color.WHITE
            textSize = 12f.sp
        }
    }
    private val bgPaint by lazy {
        Paint().apply {
            isAntiAlias = true
            color = Color.RED
        }
    }
    private var backgroundColor: Int = -1
    private var backgroundBorderColor: Int = -1
    private var backgroundBorderWidth: Int = -1
    private lateinit var textBounds: Rect
    private lateinit var bgRect: RectF
    private var verticalPadding = 0f.dp// 垂直padding
    private var horizontalPadding = 0f.dp// 水平padding
    var text: String = ""
        set(value) {
            if (field == value) {
                return
            }
            val needRequestLayout = value.length != field.length
            field = value
            visibility = if (field.isEmpty()) {
                GONE
            } else {
                textBounds = Rect().apply {
                    textPaint.getTextBounds(field, 0, field.length, this)
                }
                VISIBLE
            }
            if (needRequestLayout) {// 需要重新onMeasure、onDraw
                requestLayout()
            } else {// 需要重新onDraw
                invalidate()
            }
        }

    init {
        this.layoutParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
            Gravity.END or Gravity.TOP
        )
    }

    fun setTargetView(target: View?) {
        if (parent != null) {
            (parent as ViewGroup).removeView(this)
        }
        if (target == null) {
            return
        }
        when (target.parent) {
            is FrameLayout -> {
                (target.parent as FrameLayout).addView(this@BadgeView)
            }
            is ViewGroup -> {
                // use a new FrameLayout container for adding badge
                val parentContainer = target.parent as ViewGroup
                val groupIndex = parentContainer.indexOfChild(target)
                parentContainer.removeView(target)

                FrameLayout(context).apply {
                    layoutParams = target.layoutParams

                    target.layoutParams = FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT
                    )
                    addView(target)

                    addView(this@BadgeView)
                    parentContainer.addView(this, groupIndex)
                }
            }
            null -> {
                throw IllegalArgumentException("BadgeView setTargetView failure! parent of target is null")
            }
        }
    }

    fun setTextColor(color: Int) {
        textPaint.color = color
    }

    fun setTextSize(size: Float) {
        textPaint.textSize = size
    }

    override fun setBackgroundColor(color: Int) {
        backgroundColor = color
    }

    fun setBackgroundBorder(color: Int, width: Int) {
        backgroundBorderColor = color
        backgroundBorderWidth = width
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        verticalPadding = textPaint.fontMetrics.descent / 2f
        val w = textPaint.measureText(text).toInt()
        val h =
            (textPaint.fontMetrics.descent - textPaint.fontMetrics.ascent - textPaint.fontMetrics.leading).toInt() + (verticalPadding * 2).toInt()
        if (text.length < 2) { // 只有1个字符，就画圆
            horizontalPadding = (h - w) / 2f
            super.setMeasuredDimension(h, h)
        } else {
            horizontalPadding = textPaint.fontMetrics.descent * 2f
            super.setMeasuredDimension(w + (horizontalPadding * 2f).toInt(), h)
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        bgRect = if (backgroundBorderWidth > 0) {
            RectF(
                0f + backgroundBorderWidth,
                0f + backgroundBorderWidth,
                (width - backgroundBorderWidth).toFloat(),
                (height - backgroundBorderWidth).toFloat()
            )
        } else {
            RectF(0f, 0f, width.toFloat(), height.toFloat())
        }
    }

    override fun onDraw(canvas: Canvas) {
        if (text.isEmpty()) return
        if (backgroundBorderWidth > 0) {
            bgPaint.color = backgroundBorderColor
            bgPaint.strokeWidth = backgroundBorderWidth.toFloat()
            bgPaint.style = Paint.Style.STROKE
            canvas.drawRoundRect(bgRect, height / 2f, height / 2f, bgPaint)
        }
        bgPaint.color = backgroundColor
        bgPaint.style = Paint.Style.FILL
        canvas.drawRoundRect(bgRect, height / 2f, height / 2f, bgPaint)
        canvas.drawText(
            text,
            horizontalPadding,
            -textPaint.fontMetrics.ascent - textPaint.fontMetrics.leading + verticalPadding,
            textPaint
        )
    }

}